moveit_ros/planning/planning_request_adapter_plugins/src/fix_start_state_collision.cpp

```diff
@@ -115,37 +127,97 @@ class FixStartStateCollision : public planning_request_adapter::PlanningRequestA
+      std::vector<const moveit::core::LinkModel*> link_models;
+      link_models.resize(jmodels.size());
+      for (std::size_t i = jmodels.size() - 1; i > 0; --i)
+        link_models[i] = jmodels[i]->getChildLinkModel();
+
+      if (std::find(link_models.begin(), link_models.end(), ee_link_model) == link_models.end())
+      {  // If the end effector is a separate link, add it to the front of the vector
+        link_models.resize(jmodels.size() + 1);
+        link_models[0] = ee_link_model;
+        for (std::size_t i = jmodels.size() - 1; i > 0; --i)
+          link_models[i + 1] = jmodels[i]->getChildLinkModel();
+      }
+
+      // Store the unchanged link positions
+      EigenSTL::vector_Isometry3d original_link_poses;
+      original_link_poses.resize(link_models.size());
+      Eigen::Isometry3d new_link_pose;
+
+      for (std::size_t i = 0; i < link_models.size(); ++i)
+        original_link_poses[i] = start_state.getGlobalLinkTransform(link_models[i]);
+
+      double dist, max_dist;
+      double local_jiggle_fraction = jiggle_fraction_;
       bool found = false;
-      for (int c = 0; !found && c < sampling_attempts_; ++c)
+      for (int c = 0; !found && c < max_sampling_attempts_; ++c)
       {
+        // Change each joint slightly ("jiggle")
         for (std::size_t i = 0; !found && i < jmodels.size(); ++i)
         {
           std::vector<double> sampled_variable_values(jmodels[i]->getVariableCount());
-          const double* original_values = prefix_state->getJointPositions(jmodels[i]);
+          const double* original_values = original_state->getJointPositions(jmodels[i]);
           jmodels[i]->getVariableRandomPositionsNearBy(rng, &sampled_variable_values[0], original_values,
                                                        jmodels[i]->getMaximumExtent() * jiggle_fraction_);
           start_state.setJointPositions(jmodels[i], sampled_variable_values);
-          collision_detection::CollisionResult cres;
-          planning_scene->checkCollision(creq, cres, start_state);
-          if (!cres.collision)
+        }
+
+        // Confirm that no links move far after jiggling
+        max_dist = -1;
+        bool euclidean_distance_ok = true;
+        for (std::size_t i = 0; !found && i < link_models.size(); ++i)
+        {
+          new_link_pose = start_state.getGlobalLinkTransform(link_models[i]);
+          dist = (new_link_pose.translation() - original_link_poses[i].translation()).squaredNorm();
+
+          if (dist > max_dist)
+            max_dist = dist;
+          if (dist > max_jiggle_distance_squared_)
           {
-            found = true;
-            ROS_INFO("Found a valid state near the start state at distance %lf after %d attempts",
-                     prefix_state->distance(start_state), c);
+            // Reduce jiggle fraction if it caused excessive movement
+            local_jiggle_fraction = local_jiggle_fraction * ((max_sampling_attempts_ - 2) / max_sampling_attempts_);
```

Comment:
As these are arbitrary choices anyway, why not `max_sampling_attempts_ / (max_sampling_attempts_ + 2)`?
Could you also add a comment explaining that this is an arbitrary way to implement it and other ways might be more efficient?

I'm not really happy with this mechanism as-is, but I also do not want to burden you with implementing a more sophisticated approach because, as you say, this is an edge-case.

Reply:
Done. I don't mind taking it out either though, I doubt that it does much.